File: online-store-api/online-store-api/src/main/java/com/code81/onlinestore/service/CustomerService.java
package com.code81.onlinestore.service;

import com.code81.onlinestore.dto.customer.ChangePasswordRequest;
import com.code81.onlinestore.dto.customer.CustomerProfileUpdateRequest;
import com.code81.onlinestore.dto.customer.CustomerRegisterRequest;
import com.code81.onlinestore.dto.customer.CustomerResponse;

public interface CustomerService {

    CustomerResponse register(CustomerRegisterRequest request);

    CustomerResponse getProfile(Long customerId);

    CustomerResponse updateProfile(Long customerId, CustomerProfileUpdateRequest request);

    void changePassword(Long customerId, ChangePasswordRequest request);
}

File: code81_online_store_angular/src/app/core/services/customer.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { environment } from '../../../environments/environment';
import {
  ChangePasswordRequest,
  CustomerProfileUpdateRequest,
  CustomerRegisterRequest,
  CustomerResponse
} from '../models/customer.model';

@Injectable({ providedIn: 'root' })
export class CustomerService {
  private readonly baseUrl = `${environment.apiUrl}/customers`;

  constructor(private http: HttpClient) {}

  register(request: CustomerRegisterRequest): Observable<CustomerResponse> {
    return this.http.post<CustomerResponse>(`${this.baseUrl}/register`, request);
  }

  getMyProfile(): Observable<CustomerResponse> {
    return this.http.get<CustomerResponse>(`${this.baseUrl}/me`);
  }

  updateMyProfile(request: CustomerProfileUpdateRequest): Observable<CustomerResponse> {
    return this.http.put<CustomerResponse>(`${this.baseUrl}/me`, request);
  }

  changePassword(request: ChangePasswordRequest): Observable<void> {
    return this.http.post<void>(`${this.baseUrl}/me/change-password`, request);
  }
}
